### PR TITLE
feat: replace relative import closes #5484

### LINKED ISSUE #5484: Replace relative imports to `@`
Our current `tsconfig` allows all files compiled via TypeScript to use `@` to indicate the `root` folder.

This allows us to replace long relative imports such as `import something from '../../../../big-path/something';` to simply `import something from '@/big-path/something';`

This PR is about updating all these relative paths with `@` by doing the work manually or having an ESLint rule from `.ts/.tsx` files to suggest that and fix that automatically.

### PR DESCRIPTION
**Pull request by umairraza96**
## Description

Added a plugin for eslint `eslint-plugin-no-relative-import-paths`

### Check List

- [x] I have read the [Contributing Guidelines](https://github.com/nodejs/nodejs.org/blob/main/CONTRIBUTING.md) and made commit messages that follow the guideline.
- [x] I have run `npx turbo lint` to ensure the code follows the style guide. And run `npx turbo lint:fix` to fix the style errors if necessary.
- [x] I have run `npx turbo format` to ensure the code follows the style guide.
- [x] I have run `npx turbo test` to check if all tests are passing, and/or `npx turbo test:snapshot` to update snapshots if I created and/or updated React Components.
- [x] I've covered new added functionality with unit tests if necessary.

### CODE REVIEW

**Review by mikeesto — COMMENTED** [review]
Thanks for the PR. I think this on the right track but these changes have broken storybook. Can you please have a look at this: https://github.com/nodejs/nodejs.org/actions/runs/5499772573/jobs/10027807119?pr=5487

**Comment by umairraza96** [comment]
> Thanks for the PR. I think this on the right track but these changes have broken storybook. Can you please have a look at this: https://github.com/nodejs/nodejs.org/actions/runs/5499772573/jobs/10027807119?pr=5487

I have fixed the test issue but there is 1 snapshot test that is failing and I don't know whats causing it may be we need to record a new snapshot *need assistance* on this

**Comment by ovflowd** [comment]
> > Thanks for the PR. I think this on the right track but these changes have broken storybook. Can you please have a look at this: [nodejs/nodejs.org/actions/runs/5499772573/jobs/10027807119?pr=5487](https://github.com/nodejs/nodejs.org/actions/runs/5499772573/jobs/10027807119?pr=5487)
> 
> I have fixed the test issue but there is 1 snapshot test that is failing and I don't know whats causing it may be we need to record a new snapshot _need assistance_ on this

Can you first rebase your branch? We did a few changes on some files that are conflicting.

Now, to update the snapshots you can simply run `npm run test:storybook:snapshot`

**Comment by umairraza96** [comment]
> > > Thanks for the PR. I think this on the right track but these changes have broken storybook. Can you please have a look at this: [nodejs/nodejs.org/actions/runs/5499772573/jobs/10027807119?pr=5487](https://github.com/nodejs/nodejs.org/actions/runs/5499772573/jobs/10027807119?pr=5487)
> > 
> > 
> > I have fixed the test issue but there is 1 snapshot test that is failing and I don't know whats causing it may be we need to record a new snapshot _need assistance_ on this
> 
> Can you first rebase your branch? We did a few changes on some files that are conflicting.
> 
> Now, to update the snapshots you can simply run `npm run test:storybook:snapshot`

Could you please check the PR and let me know if anything is required from my side

**Comment by mikeesto** [comment]
Storybook looks good now, nice one

File: public/node-releases-data.json
Comment (mikeesto) [inline public/node-releases-data.json:1]: This file shouldn't have a diff. There's a pre-commit hook that perhaps hasn't run for you

File: public/blog-posts-data.json
Comment (mikeesto) [inline public/blog-posts-data.json:1]: Same with this file

File: public/node-releases-data.json
Comment (umairraza96) [reply public/node-releases-data.json:1]: I'll have a look.

File: .github/workflows/build-and-analysis.yml [hunk 1]
```diff
@@ -141,7 +141,7 @@ jobs:
         id: find-comment-id
         with:
           issue-number: ${{ github.event.number }}
-          body-includes: '<!-- __NEXTJS_BUNDLE_nodejs.org -->'
+          body-includes: '<!-- __NEXTJS_BUNDLE -->'
```
Comment (ovflowd) [inline .github/workflows/build-and-analysis.yml]: Why is this being changed?

File: .storybook/main.ts [hunk 2]
```diff
@@ -8,6 +9,22 @@ const config: StorybookConfig = {
   docs: { autodocs: 'tag' },
   staticDirs: ['../public'],
   core: { disableTelemetry: true },
+  webpackFinal: async (config, { configType }) => {
+    config.resolve!.modules = [path.resolve(__dirname, '..'), 'node_modules'];
+
+    config.resolve!.alias = {
```
Comment (ovflowd) [inline .storybook/main.ts:15]: Is this the best way of doing so? Could you share how you came to this conclusion?

File: public/blog-posts-data.json
Comment (umairraza96) [reply public/blog-posts-data.json:1]: So what i did is made a new commit with `--no-verify` just to bypass commit hook and reverted back the files to their previous state

File: .storybook/main.ts [hunk 3]
```diff
@@ -1,4 +1,5 @@
 import type { StorybookConfig } from '@storybook/nextjs';
+import path from 'path';
```
Comment (ovflowd) [inline .storybook/main.ts]:
```suggestion
import { resolve } from 'node:path';
```

File: public/node-releases-data.json
Comment (ovflowd) [reply public/node-releases-data.json:1]: @bmuenzenmeyer I wonder if the precommit hook is not being installed for reasons? Or maybe it matters how they do commit? Like which tool they use to commit?

@umairraza96 what tool are you using for commits? The hook should be installed if you run `npm install`

File: .storybook/main.ts (hunk 2 above)
Comment (umairraza96) [reply .storybook/main.ts:15]: Custom Webpack Configuration for storybook
https://storybook.js.org/docs/react/builders/webpack#import-a-custom-webpack-configuration

File: public/blog-posts-data.json
Comment (ovflowd) [reply public/blog-posts-data.json:1]: Why you had to use `--no-verify`, you're doing something probably really wrong. I should mention that once you rebase your branch with `main` you should also run `npm ci`.

Also our pre-commit hook should pretty much discard this file from the commit.

File: public/node-releases-data.json
Comment (umairraza96) [reply public/node-releases-data.json:1]: I am using vim fugitive and I think it messed something up

File: .storybook/main.ts (hunk 2 above)
Comment (ovflowd) [reply .storybook/main.ts:15]: Sure, but that's not my question. Why are you doing it like this? (Like why are you changing `config.resolve.modules?`

```tsx
webpackFinal: (config) => {
  config.resolve.alias = {
    ...config.resolve.alias,
      "@": resolve(__dirname, '..'),
  };

  return config;
},
```

Pretty much I want to understand why you're doing it this way.

File: .github/workflows/build-and-analysis.yml (hunk 1 above)
Comment (umairraza96) [reply .github/workflows/build-and-analysis.yml]: I have no clue about this one, I think it came with the rebase

File: .storybook/main.ts (hunk 2 above)
Comment (umairraza96) [reply .storybook/main.ts:15]: Because storybook doesn't support Next.js absolute paths out of the box and we have to tell it how to do relative paths through webpack 
https://github.com/storybookjs/storybook/issues/11639

File: public/node-releases-data.json
Comment (ovflowd) [reply public/node-releases-data.json:1]: Hmmm 🤔

File: .storybook/main.ts (hunk 2 above)
Comment (ovflowd) [reply .storybook/main.ts:15]: Sorry, let me rephrase my question. First, yes I got that you made this custom config because Storybook does not support the Next.js aliases.

My question was about why the config.modules, but I suppose that after looking at the linked issue, I think I get what is being done there. Still I don’t understand why you added manual aliases for each folder. On the linked issue they’ve done with a simple @ alias 🤔

Comment (ovflowd) [reply .storybook/main.ts:15]: Like this:

```tsx

webpackFinal: async config => {
    config.resolve.modules = [resolve(__dirname, '..'), 'node_modules'];
    config.resolve.alias = {
      ...config.resolve.alias,
      '@': resolve(__dirname, '../'),
    };
    return config;
  },

```

Comment (ovflowd) [reply .storybook/main.ts:15]: Also that resolve could become a simple constant on the beginning of the file for reusability sake

File: public/blog-posts-data.json
Comment (umairraza96) [reply public/blog-posts-data.json:1]: Yeah I tried with vanilla git commands and precommit hook was working fine however somehow these 2 files that @mikeesto mentioned got through  precommit hook may be because I used `vim fugitive` a neovim's plugin and it disregarded the precommit hook, since these two files have been changed I had to revert back these files to their previous version and now precommit hook was working as expected and it wouldn't let me commit those changes to files hence I used `--no-verify` option in order to forcefully commit these changes

File: .storybook/main.ts (hunk 2 above)
Comment (umairraza96) [reply .storybook/main.ts:15]: > Sorry, let me rephrase my question. First, yes I got that you made this custom config because Storybook does not support the Next.js aliases.
> 
> My question was about why the config.modules, but I suppose that after looking at the linked issue, I think I get what is being done there. Still I don’t understand why you added manual aliases for each folder. On the linked issue they’ve done with a simple @ alias thinking

I tried that but it didn't pick up and it was complaining about hooks and other components that it couldn't find so I figured these values should be hard coded and same with the modules I think it breaks the fact that it could look into node_modules for storybook

Comment (ovflowd) [reply .storybook/main.ts:15]: This definitely requires more investigation. Hard-coding this is sadly a no-go :(

File: .storybook/main.ts (hunk 3 above)
Comment (umairraza96) [reply .storybook/main.ts]: sorry I am new to this, should i sign off this commit in order to accept the change ?

Comment (ovflowd) [reply .storybook/main.ts]: Yeah, that’s fine 🙃 (you can click the button that shows on the UI to sign-off the commit)

File: .storybook/main.ts (hunk 2 above)
Comment (umairraza96) [reply .storybook/main.ts:15]: I agree with that

Comment (mikeesto) [reply .storybook/main.ts:15]: > Like this:
> 
> ```tsx
> webpackFinal: async config => {
>     config.resolve.modules = [resolve(__dirname, '..'), 'node_modules'];
>     config.resolve.alias = {
>       ...config.resolve.alias,
>       '@': resolve(__dirname, '../'),
>     };
>     return config;
>   },
> ```

FWIW, this worked fine for me locally. I was also able to run storybook build and serve it.

Comment (umairraza96) [reply .storybook/main.ts:15]: I will look into this once more if its running for you then I must be doing something wrong.

Comment (umairraza96) [reply .storybook/main.ts:15]: @mikeesto It is failing for me in local as well

image

Comment (ovflowd) [reply .storybook/main.ts:15]: For me, it looks like a cache issue. The approach above also works for me.

You can test that it works in CI by committing the example. Or @mikeesto can take-over the PR, at least with this change. Maybe he tweaked it a little bit.

Comment (umairraza96) [reply .storybook/main.ts:15]: got it I was confused with the `/` turns out `/` was causing the issue with me and I thought that `/` was also required for directory mapping

Comment (ovflowd) [reply .storybook/main.ts:15]: Yeah the `@` is not required at all 🙈 actually should not even be there! What we're aliasing is `@` and not `@/` because `@/` is already two different directories.

Glad to hear you solved it!

**Comment by ovflowd** [comment]
FYI rebasing seems to be needed!

File: .github/workflows/build-and-analysis.yml (hunk 1 above)
Comment (ovflowd) [reply .github/workflows/build-and-analysis.yml]: So you can discard these changes, either by `git checkout main -- .github/workflows/build-and-analysis.yml` or by other means.

Please give double-check when rebasing that you're not accidentally reverting things or introducing bugs! Even more on such a big PR that introduces so many file changes 🙇

Comment (umairraza96) [reply .github/workflows/build-and-analysis.yml:144]: @ovflowd should I add this back?

Comment (ovflowd) [reply .github/workflows/build-and-analysis.yml]: If you run the command I’ve shown it should fix the file 🙃 even if you commit the fix in the end all the changes here get squashed.

Another alternative is to discard this branch (delete it locally, checkout main and “git pull”, then create the branch again and simply add your eslint changes and storybook changes and run the eslint fix command again 🙃

That probably would be easier.

Comment (umairraza96) [reply .github/workflows/build-and-analysis.yml]: Reverted this file using your git command 
https://github.com/nodejs/nodejs.org/pull/5487/commits/56e7951cdbd8040816df1a47d56f85973f806483

**Comment by ovflowd** [comment]
@umairraza96 could we do the rebasing or recreation once more? And then I'll prioritise getting this merged so that you don't need to worry about further rebases/updates! 🙇

File: .github/workflows/build-and-analysis.yml [hunk 4]
```diff
@@ -46,7 +46,7 @@ jobs:
             ~/.npm
             .next/cache
             node_modules/.cache
-          key: cache-${{ hashFiles('package-lock.json') }}
```
Comment (ovflowd) [inline .github/workflows/build-and-analysis.yml:49]: Unrelated changes again over here :S

File: public/node-releases-data.json [hunk 5]
```diff
@@ -1 +1 @@
-[]
```
Comment (ovflowd) [inline public/node-releases-data.json:1]: Also unrelated changes here

Comment (AugustinMauroy) [reply public/node-releases-data.json:1]: Why we don’t put this file in gitinignore ?

Comment (ovflowd) [reply public/node-releases-data.json:1]: This was already discussed before. And the file should not be on gitignore. (Also adding it right now would have 0 effects)

**Comment by umairraza96** [comment]
@ovflowd I created a new branch https://github.com/nodejs/nodejs.org/pull/5513 which closes #5487

**Comment by ovflowd** [comment]
I see, you could simply have deleted this one, rebranch from main and force push, so that you wouldn’t need to create a new PR 🙈

But that’s okay, let’s close this one!

**Comment by mikeesto** [comment]
Closing in favor of https://github.com/nodejs/nodejs.org/pull/5513
